### ISSUE TITLE
Guaranteed copy elision for potentially-overlapping non-static data members

### ISSUE THREAD

**Issue by jwakely** [itanium-cxx-abi/cxx-abi#107]
GCC 11 and Clang 11 disagree whether this C++17 program is ill-formed:

```cpp
template<typename T>
  struct tuple
  {
    char c1 = ~0;
    [[no_unique_address]] T t;
    char c2 = ~0;

    template<typename U>
      tuple(U&& u)
      : t(static_cast<U&&>(u))
      { }
  };

struct Bar
{
    Bar(const Bar&) = delete;
};

struct Foo
{
    operator Bar() const { return {}; }
};

int main()
{
  tuple<Bar> t{ Foo{} };
  static_assert(sizeof(tuple<Bar>) == 2);
  if (t.c1 == 0 || t.c2 == 0)
    throw 1;
}
```

GCC says this is ill-formed because it disables copy elision for the potentially-overlapping subobject, which is consistent with its treatment of empty base classes. Clang compiles it without error. Both compilers refuse to elide the copy for an empty base class.

It would be useful if the ABI clarified whether reusing the tail padding here is allowed/required. If it should be reused, then implementations either need to refuse to do guaranteed elision (despite the C++ standard requiring it) or need to avoid storing to the tail padding of objects which might be potentially-overlapping subobjects.

**Comment by zygoloid** [comment]
An implementation already can't store to the tail padding of an object if that object might be a base subobject of some other object. Given that any object of class type of unknown provenance might be the `this` pointer in a constructor for a base class subobject, I think we currently have the following de facto rule:

You may store to the tail padding of an object if:
1) It is POD for the purpose of layout (tail padding is not reusable), or
2) You know where it came from, and it's not a base class subobject, or
3) It's the return value in a function.

(3) only works if we can assume that a return value is never a base class subobject, which is at least one reason why we disable copy elision for base class subobjects. I think we could refine the above rule to:

You may store to the tail padding of an object if:
1) It is POD for the purpose of layout (tail padding is not reusable), or
2) You know where it came from, and it's not a potentially-overlapping object.

In particular: never assume we know where the return value of a function came from. And in-double-particular: when NRVO is applied to a local variable, it fails the "you know where it came from" test.

This does render existing implementations non-conforming: for example, both GCC and Clang store to the tail padding of an NRVO object here: https://godbolt.org/z/dhn6fT -- but in at least Clang's case this existing behavior can be easily extended into a miscompile: https://godbolt.org/z/3GKsa7.

**Comment by jicama** [comment]
I'd think [[no_unique_address]] makes the example a C++20 program, not C++17?

I raised the issue for bases on the ABI list in 2016, but the list archive seems to be offline now.  The subject was "Guaranteed copy elision and tail padding".  That discussion ended with John McCall's suggestion that we disable copy elision for bases, which is what G++ implements.  And Clang, apparently?  I've now extended that to [[no_unique_address]] members because the problem with padding is the same.

The [2018 Core discussion](http://lists.isocpp.org/core/2018/02/4003.php) didn't come to a conclusion.  A recent revival of that thread pointed out [CWG2403](http://wg21.link/cwg2403), which has no proposal.

Since C++17 mandatory copy elision is a language rule, I think we need a fix in the language, not just the ABI.  My theory has been that to avoid depending on a particular ABI, we should always do a copy when initializing a base or potentially-overlapping subobject from a function call.  The other plausible resolution is to require a copy only when initializing a virtual base, and follow your suggestion to make copy elision work in other cases.

One obstacle to that is the [CWG694](https://wiki.edg.com/pub/Wg21pittsburgh/CoreWorkingGroup/proposed_resolution_core-694.html) rule that zero-initialization includes zeroing padding bits; this conflicts with your suggestion, as on returning a value-initialized object we would be required both to initialize and not initialize the padding.

Perhaps the CWG694 padding initialization should only apply to trivially-copyable types, since those are the types for which it makes any sense for the user to touch the padding.

But as you say, your  this is an ABI change, that seems likely to cause runtime failures when new code calls old code that clobbers the tail padding of the return slot.  Doing the copy seems like the solution that isn't an ABI break.

**Comment by rjmccall** [comment]
I'll see if I can get the full mailing-list archives from Mentor Embedded.  Some of them still seem to be up, but the archive index pages aren't.

**Comment by rjmccall** [comment]
Wait.  My understanding is that mandatory copy elision comes down to this bullet in [dcl.init]p17 (taken from N4820):

> If the initializer expression is a prvalue and the cv-unqualified version of the source type is the same class as the class of the destination, the initializer expression is used to initialize the destination object.

This rule is unimplementable if it includes the initialization of base subobjects because the source type may have virtual bases which will already have been separately initialized in the destination complete object.

**Comment by zygoloid** [comment]
> This rule is unimplementable if it includes the initialization of base subobjects because the source type may have virtual bases which will already have been separately initialized in the destination complete object.

Yes, this is a known bug in the language rules, and the tentative solution is that guaranteed copy elision doesn't apply for a base class initializer.

> Perhaps the CWG694 padding initialization should only apply to trivially-copyable types, since those are the types for which it makes any sense for the user to touch the padding.

That makes some sense, but Clang has support for ABIs whose struct layout rule reuses the tail padding for even trivially-copyable types, so it's probably not enough.

> But as you say, this is an ABI change, that seems likely to cause runtime failures when new code calls old code that clobbers the tail padding of the return slot. Doing the copy seems like the solution that isn't an ABI break.

I think we need to perform the copy in the case where the `[[no_unique_address]]` subobject is empty, but it should at least be cheap in that case. When it's non-empty, we only need the copy when an earlier-initialized subobject is laid out later, which I think can only happen when a virtual base reuses the tail padding. Maybe we could split this hair, but ... I think the cleanest and simplest language rule would be that you don't get copy-elision when initializing any potentially-overlapping subobject.

**Comment by jicama** [comment]
This also comes up in the context of C++17 aggregate base initialization.

```
// Empty base
struct base {
  base(void) {};
  base(base &&) = delete;
};

class foo : public base { };

foo c1{ };         // IMO needs to be OK (GCC currently rejects)
foo c2 { {} };     // equivalent in C++17
foo c3 { base{} }; // not OK?
```

It seems a bit unfortunate that making the type of the initializer explicit changes the behavior so much; theoretically we could recognize that this is a constructor call and use it directly, tweaking it to call the base constructor instead of derived.

More broadly we could say that you *do* get copy elision when initializing a potentially-overlapping subobject from a constructor call; this could happen in mem-initializers as well.  If calling the copy constructor is safe, calling another constructor should also be safe.  This would mean more guaranteed copy elision, but heightens the difference between constructors and return by value.  I don't have a strong opinion either way.

**Comment by rjmccall** [comment]
That's an interesting idea; it is definitely technically feasible to treat this as a direct initialization of the base and to just call the base-subobject constructor, and that probably best matches programmer intent.  We would definitely need the standard's blessing for it, though.

**Comment by jicama** [comment]
> We would definitely need the standard's blessing for it, though.

Well, until CWG2403 is resolved, the standard still requires that we always do direct initialization; that's impossible in the current ABI, but this can get us closer.

**Comment by frederick-vs-ja** [comment]
I don't think storing to the padding during initialization is problematic unless a virtual base subobject is touched. When a subobject is initialized later, IIUC it's safe to place it into the tail padding of a previously initialized subobject.

I think we should introduce a new UB to explicitly permit (if not to effectively require) such storing:
> Some operations are described as _writing to non-occupied bytes_ of a potentially-overlapping suboject during its initialization. If an operation writes to non-occupied bytes of a potentially-overlapping subobject `o` during its initialization, and
> - the innermost object that contains `o` and is not a potentially-overlapping subobject contains a virtual base subobject, and
> - the completion of the initialization of the virtual base subobject is sequenced before the initialization that operation,
>
> the behavior is undefined.

These operations includes:
- zero-initialization,
- materialization of the result of `std::bit_cast`, and
- some forms of object representation rewriting, including `std::memset` (see also [this](https://stackoverflow.com/questions/54102890/class-using-virtual-inheritance-seems-to-allow-a-base-class-constructor-to-overw)). Unfortunately, the validity of this kind of operations seems underspecified now even if no potentially-overlapping suboject is involved. Perhaps [P1839](https://wg21.link/p1839) will solve the problem.
